File: IAIso-v5.0/core/iaiso-swift/Sources/IAIsoPolicy/Aggregator.swift
import Foundation

/// A coordinator aggregation strategy.
public protocol Aggregator: Sendable {
    /// Wire-format aggregator name.
    var name: String { get }

    /// Compute the aggregate from per-execution pressures.
    func aggregate(_ pressures: [String: Double]) -> Double
}

/// Sum of per-execution pressures.
public struct SumAggregator: Aggregator {
    public init() {}
    public var name: String { "sum" }
    public func aggregate(_ pressures: [String: Double]) -> Double {
        return pressures.values.reduce(0, +)
    }
}

/// Arithmetic mean of per-execution pressures.
public struct MeanAggregator: Aggregator {
    public init() {}
    public var name: String { "mean" }
    public func aggregate(_ pressures: [String: Double]) -> Double {
        if pressures.isEmpty { return 0 }
        return pressures.values.reduce(0, +) / Double(pressures.count)
    }
}

/// Maximum of per-execution pressures.
public struct MaxAggregator: Aggregator {
    public init() {}
    public var name: String { "max" }
    public func aggregate(_ pressures: [String: Double]) -> Double {
        return pressures.values.max() ?? 0
    }
}

/// Weighted sum of per-execution pressures.
public struct WeightedSumAggregator: Aggregator {
    public let weights: [String: Double]
    public let defaultWeight: Double

    public init(weights: [String: Double], defaultWeight: Double = 1.0) {
        self.weights = weights
        self.defaultWeight = defaultWeight
    }

    public var name: String { "weighted_sum" }

    public func aggregate(_ pressures: [String: Double]) -> Double {
        var total = 0.0
        for (k, v) in pressures {
            let w = weights[k] ?? defaultWeight
            total += w * v
        }
        return total
    }
}

File: IAIso-v5.0/core/iaiso-ruby/lib/iaiso/policy/aggregator.rb
# frozen_string_literal: true

module IAIso
  module Policy
    # Aggregator interface (documentation; duck-typed).
    module Aggregator
      def name; raise NotImplementedError; end
      def aggregate(pressures); raise NotImplementedError; end
    end

    # Sum of per-execution pressures.
    class SumAggregator
      def name = "sum"
      def aggregate(pressures)
        pressures.values.sum.to_f
      end
    end

    # Arithmetic mean.
    class MeanAggregator
      def name = "mean"
      def aggregate(pressures)
        return 0.0 if pressures.empty?
        pressures.values.sum.to_f / pressures.size
      end
    end

    # Maximum.
    class MaxAggregator
      def name = "max"
      def aggregate(pressures)
        return 0.0 if pressures.empty?
        pressures.values.max.to_f
      end
    end

    # Weighted sum.
    class WeightedSumAggregator
      attr_reader :weights, :default_weight

      def initialize(weights:, default_weight: 1.0)
        @weights = weights || {}
        @default_weight = default_weight.to_f
      end

      def name = "weighted_sum"

      def aggregate(pressures)
        total = 0.0
        pressures.each do |k, v|
          w = @weights[k.to_s] || @default_weight
          total += w * v
        end
        total
      end
    end
  end
end
